keps/sig-node/20191205-container-streaming-requests.md

```diff
@@ -0,0 +1,211 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+There are several features and flags for handling container streaming requests
+(exec/attach/port-forward). Having so many different ways of handling these requests unnecessarily
+complicates configuration, and opens the cluster to potential security risks. This is a proposal to
+consolidate the current options into a single supported configuration. Specifically, it proposes:
+
+1. Deprecate the Kubelet's `--redirect-container-streaming` feature
+2. Deprecate the `StreamingProxyRedirects` feature
+
+## Motivation
+
+The `StreamingProxyRedirects` feature was originally added as part of the migration the [Container
+Runtime Interface][]. One of the initial goals was to bypass the Kubelet for these long-running
+requests in order to better account the request overhead. However, the apiserver is still
+responsible for proxying these requests, so the gains of bypassing the Kubelet are
+marginal. Furthermore, the redirect handling significantly complicates the apiserver stream
+handling, and opens it up to [potential SSRF attacks][], so we added the `ValidateProxyRedirects`
+feature to mitigate those.
+
+When we went to add mTLS support to the runtime streaming server, we realized getting the right
+certs established between the apiserver and the runtime was complicated, but the apiserver already
+had an mTLS connection to the Kubelet. So, we added the ability for the Kubelet to connect the
+stream locally and proxy the connection back to the apiserver. This setup is more secure, removes
+the need for `StreamingProxyRedirects` and `ValidateProxyRedirects`, but adds the Kubelet back into
+the long-running request chain.
+
+`StreamingProxyRedirects` and `ValidateProxyRedirects` are currently in Beta, and should either be
+graduated or removed.
+
+[Container Runtime Interface]: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-runtime-interface-v1.md
+[potential SSRF attacks]: https://github.com/kubernetes/kubernetes/issues/85867
+
+### Goals
+
+1. Simplify secure container runtime setup and operation
+2. Remove conformance dependencies on beta features
+
+### Non-Goals
+
+Design a high-bandwidth low-latency production ready streaming channel to containers. Such a system
+should go through an ingress point other than the apiserver, and could be implemented (or at least
+prototyped) out of tree by leveraging the new [ephemeral containers
+feature](20190212-ephemeral-containers.md).
```

Comment:
I don't see how ephemeral containers out of tree can replace exec?

Reply:
I can see an argument that the ability to run an ephemeral container in a pid namespace of a particular container in a pod spec is similar...

Reply:
It would be nice to preserve the exec functionality so it transparently created/deleted the ephemeral container; using this method.

Will the functionality of exec and attach (maybe less so for port-forward) be preserved?

Reply:
To clarify, the current exec/attach/port-forward behavior will continue to be supported, but with the extra network hop through the Kubelet. This is how it currently works on GKE, and I think the extra hop is negligible for the majority of use cases.

The suggestion here is specifically for when a high-bandwidth and/or low latency production system is needed. With the current approach, the apiserver is likely to be a bandwidth bottleneck. In this case, connecting directly through ingress to a server (e.g. SSH) in the pod would be preferable. This could be brought up on the fly with an ephemeral container (as described here), or it could be built into the target container for common tasks.

Does that make sense?

Reply:
Thanks for the explanation.